platform/socket.hpp

```diff
@@ -1,31 +1,78 @@
 #pragma once
 
+#include "std/mutex.hpp"
 #include "std/string.hpp"
 #include "std/target_os.hpp"
-
-#if defined(OMIM_OS_IPHONE) || defined(OMIM_OS_MAC)
-@class SocketImpl;
-#else
-class SocketImpl;
-#endif
+#include "std/vector.hpp"
 
 namespace platform
 {
 class Socket
 {
 public:
-  Socket();
-  ~Socket();
+  virtual ~Socket() {}
+
+  virtual bool Open(string const & host, uint16_t port) = 0;
+  virtual void Close() = 0;
+
+  virtual bool Read(uint8_t * data, uint32_t count) = 0;
+  virtual bool Write(uint8_t const * data, uint32_t count) = 0;
+
+  virtual void SetTimeout(uint32_t milliseconds) = 0;
+};
+
+class PlatformSocket final : public Socket
+{
+  PlatformSocket();
+  virtual ~PlatformSocket();
+  virtual bool Open(string const & host, uint16_t port) override;
+  virtual void Close() override;
+  virtual bool Read(uint8_t * data, uint32_t count) override;
+  virtual bool Write(uint8_t const * data, uint32_t count) override;
+  virtual void SetTimeout(uint32_t milliseconds) override;
```

Comment:
@dobriy-eeh Would be great if you documents this interface and its methods. Especially will be useful the contract rules and information about thread safety.